Set home directory parent tree permissions to 0755

TRPB opened the pull request:
## Description of changes


If the user's home directory is set using this code:

```nix
users.users.tom = {
    home = "/users/tom";
    createHome = true;
};
```

Then the directory structure gets created like this:

```
0700 root /users 
0700 tom /users/tom
```

The issue is due to `make_path($u->{home}, { mode => oct($u->{homeMode}) })` and `homeMode` defaulting to `0700` causes `/users` to be inaccesble to the `tom` user. In this example `cd ~` (or `cd /users/tom)` gives `Permission denied`.

While `0700` absolutely makes sense for the actual home directory and the default permission for `homeMode` it does not make sense to apply that mode to the parent folder. We should use `0755` which is the standard permission applied to directories like `/var`, `/etc`, `/usr`, `/dev` etc so that the user can view the contents and navigate into the directory.

`homeMode` is still used two lines later to configure the actual home directory.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jtojnar commented on nixos/modules/config/update-users-groups.pl:
We might also want to add a check for this to `nixos/tests/user-home-mode.nix`. (Run with `nix-build -A nixosTests.user-home-mode`.)

TRPB replied:
Thanks! I'd assumed feature branches were squash merged into master but I've squashed them and added the test.

Nice simple test structure too, the more I use Nix the more I enjoy it!

jtojnar replied:
We allow multiple related "atomic" commits per PR so we typically do not use squash merges to preserve the commits for the purposes of `git blame` and other git archaeology.

We can a squash merge if needed but it is easy to forget so it is better to leave it to the author – they can usually also come up with a more suitable commit message.

jtojnar reviewed (commented):
Thanks, makes sense to me.

I guess there might be some cases where the home directory is part of some service-specific directory tree where 700 might be desirable but 755 is probably more suitable in more commonly.

Could you also please squash the two commits since the second one is just a fixup, and include the reasoning in the commit message so that it is easily accessible with `git blame` and in case GitHub is unavailable?

TRPB committed:
Set home directory parent tree permissions to 0755

When the user's home directory is created using `createHome` e.g.

```
users.users.alice = {
    home = "/users/alice";
    createHome = true;
};
```

The `/users` directory was created with the same permissions as `/users/alice`, `0700` by default.

The parent directory `/users` permissions results in `createHome` creating a home directory that is inaccessible to the user:

```
$ su alice
$ cd /user/alice
cd: permission denied: /users/alice
```

The underlying cause is `make_path($u->{home}, { mode => oct($u->{homeMode}) })` which sets, in the example above`, `/users` to `0700`. Instead it should be `0755` like other system directories `/var`, `/dev`, etc.

bjornfor reviewed (approved):
LGTM. Please prefix the commit message with `nixos: `.

7c6f434c commented:
Seems reasonable, also in context of some of the permission discussions (but this is not a part of NixOS that I keep reusing after stopping to use NixOS mainline)

fedeinthemix reviewed (commented):
In a multi user setting being able to list the content of /home allows you to discover names of other users. This gives you one more piece of information to try to manipulate the system. To prevent this while allowing access to the home directory we can use 751.

Despite the above, I agree that in most settings 755 is more sensible, but making this permission configurable would cover more use scenarios.

jtojnar commented:
Thanks. We can add an additional option when someone needs it.

7c6f434c commented:
As far as I understand, if `/home/` exists already and has good permissions, nothing is done, so the option won't provide e.g. rollbacks anyway (and then its benefits over `chmod` are reduced)
